core/src/main/scala/org/ensime/indexer/domain.scala

```diff
@@ -47,27 +47,34 @@ case class ClassName(pack: PackageName, name: String) extends FullyQualifiedName
     if (pack.path.isEmpty) name
     else ClassName.cleanupPackage(pack.fqnString + "." + name)
 
-  def internalString =
-    "L" + (if (pack.path.isEmpty) name else pack.path.mkString("/") + "/" + name) + ";"
+  def internalString = name match {
```

Comment:
this may be functionally correct but it is going to involve a *lot* of unnecessary string comparison. This code is on the critical path during indexing.

Can we perhaps introduce a new `trait` with `ClassName` and `Primitive` as implementations? Might require refactoring elsewhere to use the new trait instead of `ClassName`.